src/MagicWormhole/Internal/ClientProtocol.hs

```diff
@@ -88,35 +104,61 @@ newtype SessionKey = SessionKey ByteString
 
 -- | Exchange SPAKE2 keys with a Magic Wormhole peer.
 pakeExchange :: Peer.Connection -> Spake2.Password -> IO (Either Error SessionKey)
-pakeExchange connection password = do
-  let protocol = wormholeSpakeProtocol (Peer.appID connection)
+pakeExchange conn password = do
+  let protocol = wormholeSpakeProtocol (Peer.appID conn)
   bimap ProtocolError SessionKey <$> Spake2.spake2Exchange protocol password sendPakeMessage (atomically receivePakeMessage)
   where
-    sendPakeMessage = Peer.send connection Messages.PakePhase . spakeBytesToMessageBody
+    sendPakeMessage = Peer.send conn Messages.PakePhase . spakeBytesToMessageBody
     receivePakeMessage  = do
       -- XXX: This is kind of a fun approach, but it means that everyone else
       -- has to promise that they *don't* consume pake messages.
-      msg <- Peer.receive connection
+      msg <- Peer.receive conn
       unless (Messages.phase msg == Messages.PakePhase) retry
       pure $ messageBodyToSpakeBytes (Messages.body msg)
 
 -- | Exchange version information with a Magic Wormhole peer.
 --
 -- Obtain the 'SessionKey' from 'pakeExchange'.
 versionExchange :: Peer.Connection -> SessionKey -> IO (Either Error Versions)
-versionExchange connection key = do
-  (_, theirVersions) <- concurrently sendVersion receiveVersion
+versionExchange conn key = do
+  (_, theirVersions) <- concurrently sendVersion (atomically receiveVersion)
```

Comment:
I added `atomically` for `receiveVersion` because the type of `receiveVersion` changed from `IO (Either Error Versions)` to `STM (Either Error Versions)`. `atomically` is how you actually run an `STM` transaction.

Previously, `receiveVersion` called `receiveEncrypted` and then parsed the results. The type of `receiveEncrypted` has now changed from:

```haskell
receiveEncrypted :: Peer.Connection -> SessionKey -> IO (Either Error PlainText)
```

to 

```haskell
receiveEncrypted :: Peer.Connection -> SessionKey -> STM (Either Error (Messages.Phase, PlainText))
```

The `IO` -> `STM` part of that change is tied to this change of code:

```diff
-  message <- atomically $ Peer.receive connection
+  message <- Peer.receive conn
```

Essentially, all I did was expand the size of the transaction. I don't recall the specific motivation behind this change, but in general I try to push `IO` to the outermost layers of a program.

As for `sendEncrypted`, you're right that it doesn't use `STM` or `atomically`. When we send something, we just send it straight to the wire. There's no internal book-keeping—no state to track. STM works by retrying when there's a transaction conflict, so it doesn't make sense to do IO "atomically" in that sense.

We have to do book-keeping for receiving because messages can arrive out of order. 

The type of `receiveVersion` is `STM (Either Error Versions)`. The type of `sendVersion` is `IO ()`. (I confirmed these using an in-Emacs helper, but could put them in explicitly.)

`WS.sendBinaryData` does not interact with our STM variables. You can infer this from the chain of type signatures: we can't pass `sendBinaryData` a `TVar` (or other STM variable), and the `WS.Connection` type doesn't have any STM variables as input either. It *might* do some STM stuff internally, but we don't have to care about that. 

`lift` merely turns an `STM a` (which is the type of `retry` and hence `unless (phase == Messages.VersionPhase) retry`) into an `ExceptT Error STM a`. It's there because everything within a `do` block has to have the same monad, and `ExceptT Error STM` is a different monad to `STM`.

The `retry` (together with the guard) says that if we have received a message which is not the version phase message then abort the transaction (i.e. put it back on the `TQueue`) and wait until the queue changes.

----

Stepping back a bit, STM is all about making transactional changes to in-memory variables. An `STM a` value represents a thing that expects to run inside a single transaction. STM values can be composed, with the resulting composition being a thing that expects to run inside a single transaction. 

What happens inside an STM action *doesn't take effect* until the action is run. You do that with `atomically`. The signatures communicate this: only `IO` can actually mutate state.

Similarly, if you want to do something STM-y and then do some IO, and then do something else STM-y, you *cannot* do that inside one transaction. The type `STM a` forbids you from embedding an IO action. You can only do pure things or other STM things. This is necessary because STM works by retrying when there's a conflict. IO actions can't reliably be retried.

I hope this helps.